File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        n = len(nums);
        pre = 1
        ans = [1]*n;
        for i in range(n):
            ans[i] = pre;
            pre *=nums[i];
        suf = 1;
        for i in range(n-1,-1,-1):
            ans[i]*=suf;
            suf *=nums[i];
        return ans;

        """
        :type nums: List[int]
        :rtype: List[int]
        """
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        int pre = 1;
        for(int i=0;i<n;i++){
            ans[i] = pre;
            pre = pre*nums[i];
        }
        int suf = 1;
        for(int i=n-1;i>=0;i--){
            ans[i] = ans[i]*suf;
            suf = suf*nums[i];
        }
        return ans;
        
    }
}